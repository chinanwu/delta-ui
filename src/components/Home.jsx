import React, { useCallback, useEffect, useState } from 'react';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { postFetch } from '../functions/FetchFunctions';
import generateGameUrl from '../functions/generateGameUrl';
import getThemeClassname from '../functions/getThemeClassname';
import hasValidCharacters from '../functions/hasValidCharacters';
import { applyFrom, applyTo } from '../thunk/GameThunk.jsx';

import './Home.less';

export const Home = ({ dark, onChangeFrom, onChangeTo }) => {
	const [gameUrl, setGameUrl] = useState(generateGameUrl);
	const [error, setError] = useState(null);
	const [redirect, setRedirect] = useState(false);

	useEffect(() => {
		document.title = 'Home - Delta';
	}, []);

	const handleChange = useCallback(
		event => {
			if (event && event.target) {
				if (event.target.value) {
					const url = event.target.value;
					setGameUrl(url);
					hasValidCharacters(url)
						? setError(null)
						: setError('Game URL must only contain letters');
				} else {
					setGameUrl('');
					setError('Game URL cannot be empty');
				}
			}
		},
		[setGameUrl, setError]
	);

	const handleCreateClick = useCallback(() => {
		if (!error) {
			postFetch(
				'http://localhost:5000/api/v1/games/new',
				JSON.stringify({ url: gameUrl })
			).then(res => {
				res.success
					? setRedirect(true)
					: setError(
							`Unable to create new game with url: ${gameUrl}. Please try again`
					  );
			});
		}
	}, [error, gameUrl, setRedirect, setError]);

	return redirect ? (
		<Redirect to={`/${gameUrl}`} />
	) : (
		<div className={getThemeClassname('Home', dark)}>
			<h1 className="Home__header">Delta</h1>
			<>
				<div className="Home__gameCreate">
					<div className="Home__game">
						<input
							id="homeInput"
							className={`Home__input${error ? ' Home__input--error' : ''}`}
							type="text"
							name="gameUrl"
							value={gameUrl}
							onChange={handleChange}
						/>
						<div className="Home__btnContainer">
							<button
								id="homeCreateBtn"
								type="submit"
								className={`Home__btn${error ? ' Home__btn--error' : ''}`}
								disabled={!!error}
								onClick={handleCreateClick}
							>
								Create
							</button>
						</div>
					</div>
					<p className="Home__error" aria-hidden={!error}>
						{error}
					</p>
				</div>
				<h2>Rules</h2>
				<p className="Home__rules">
					In each game, you are given two words, a "from" word and a "to" word.
					Starting from the "from" word, one letter must changed at a time until
					the word becomes the "to" word. However, each time a letter is swapped
					out for another, the resulting new word must still be a valid
					four-letter word in the English language.
				</p>
				<h2>Example</h2>
				<p>From: "heat" -> To: "cold"</p>
				<p className="Home__example--p">
					heat -> hea<strong className="Home__example--emphasis">d</strong> ->
					he<strong className="Home__example--emphasis">l</strong>d -> h
					<strong className="Home__example--emphasis">o</strong>ld ->{' '}
					<strong className="Home__example--emphasis">c</strong>old
				</p>
				<p className="Home__example--p">Or alternatively,</p>
				<p className="Home__example--p">
					heat -> <strong className="Home__example--emphasis">m</strong>eat -> m
					<strong className="Home__example--emphasis">o</strong>at -> moa
					<strong className="Home__example--emphasis">n</strong> -> mo
					<strong className="Home__example--emphasis">o</strong>n -> moo
					<strong className="Home__example--emphasis">d</strong> -> mo
					<strong className="Home__example--emphasis">l</strong>d ->{' '}
					<strong className="Home__example--emphasis">c</strong>old
				</p>
				<p className="Home__example--p">
					For most word pairs, there are a multitude of possible answers!
				</p>
			</>
			<div
				className={getThemeClassname('Home__footer', dark)}
				aria-label="Footer"
			>
				Made by <a href="https://www.github.com/chinanwu">Chin-An Wu</a>
			</div>
		</div>
	);
};

Home.propTypes = {
	dark: PropTypes.bool,
};

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

const mapDispatchToProps = {
	onChangeFrom: applyFrom,
	onChangeTo: applyTo,
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Home);
