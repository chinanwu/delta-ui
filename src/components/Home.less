@import '../style/colours';
@import '../style/button';

.Home {
	display: flex;
	flex-direction: column;
	//justify-content: center;
	//align-items: center;
	height: 100%;
}

.Home--dark {
	@dark();
}

.Home__header {
	display: flex;
	justify-content: center;
	margin-top: 3rem;
	font-size: 3.6rem;
}

.Home__btns {
	display: flex;
	flex-direction: column;
	align-items: center;
}

.Home__btn {
	@medBtn();
	width: 15rem;
	padding-top: 1.2rem;
	padding-bottom: 1.2rem;
	font-size: 2.6rem;
}

.Home__dailyBtn {
	width: 27rem;
}

.Home__leaderboard {
	text-align: center;
	display: flex;
	flex-direction: column;
	align-items: center;
}

.Home__leaderboard--dark {
	@dark();
}

.Home__leaderboardBtn {
	@medBtn();
}

.Home__leaderboardContent {
	transition: max-height 1s;
	max-height: 0;
	width: 60%;
	overflow: hidden;
}

.Home__leaderboardContent--expanded {
	max-height: 50rem;
}

.Home--centre {
	display: flex;
	align-items: center;
	flex-direction: column;
}

.Home--centre--dark {
	@dark();
}

.Home__dailyExplanation,
.Home__rules,
.Home__example--p {
	text-align: center;
}

.Home__dailyExplanation,
.Home__rules {
	width: 60%;
}

.Home__example--p {
	margin-top: 0;
}

.Home__example--emphasis {
	color: red;
}

.Home__example--emphasis--dark {
	color: #ff8a8a;
}

.Home__footer {
	position: absolute;
	bottom: 0;
	text-align: center;
	width: 100%;
}

.Home__footer--dark {
	@dark();
}

.Home__footer--dark a {
	color: white;
}

.Home__acksBtn--dark {
	color: white;
}

.Home__acksBtn {
	border: none;
	background-color: initial;
	font-size: 1.6rem;
	cursor: pointer;
	text-decoration: underline;
	margin-bottom: 0.5rem;
}

.Home__modal {
	position: fixed;
	top: 0;
	left: 0;
	bottom: 0;
	right: 0;
	height: 100%;
	width: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	background-color: @modalBg;
	box-shadow: @box-shadow-24;
}

.Home__modalContent {
	position: relative;
	width: 30rem;
	padding: 4rem;
	background-color: white;
}

.Home__modalContent--dark {
	background-color: #1b1b1b;
	color: #eeeeee;
}

.Home__closeAcksBtn {
	position: absolute;
	top: 0;
	right: 0;
	font-size: 1.6rem;
	background-color: lightgrey;
	padding: 0.5rem 1rem;
	border: none;
	cursor: pointer;
}

.Home__closeAcksBtn--dark {
	background-color: #4d4d4d;
	color: #eeeeee;
}

.Home__modalHeader {
	text-align: center;
}

.Home__modalAck {
	text-align: center;
}

@media only screen and (max-height: 895px) {
	.Home__footer--relative {
		position: initial;
	}
}

@media only screen and (max-height: 850px) {
	.Home__footer {
		position: initial;
	}
}

@media only screen and (max-width: 830px) {
	.Home__dailyExplanation,
	.Home__rules {
		width: 80%;
	}

	.Home__leaderboardContent {
		width: 80%;
	}
}

@media only screen and (max-width: 450px) {
	.Home {
		height: auto;
	}

	.Home p {
		font-size: 125%;
	}

	.Home__btns {
		padding-left: 1rem;
		padding-right: 1rem;
	}

	.Home__btnContainer {
		width: 100%;
	}

	.Home__btn {
		width: 100%;
		font-size: 3rem;
		margin: 1rem 0;
	}

	.Home__dailyExplanation,
	.Home__rules {
		width: initial;
	}

	.Home__leaderboardContent {
		width: initial;
	}

	.Home__modalContent {
		padding: 3rem;
		margin: 4rem;
		font-size: 125%;
	}
}

@media only screen and (max-width: 375px) {
	.Home__header {
		margin-top: 0;
	}
}
