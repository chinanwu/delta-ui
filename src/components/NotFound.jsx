import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import getThemeClassname from '../functions/getThemeClassname';

import withTitle from './HOC/withTitle.jsx';
import ThemeToggle from './ThemeToggle.jsx';

import './NotFound.less';

export const NotFound = ({ dark }) => (
	<div className={getThemeClassname('NotFound', dark)}>
		<ThemeToggle />
		<h1>404 - Page Not Found!</h1>
		<h2 className="NotFound__linkContainer">
			<span className="NotFound__linkWords">Perhaps you'd like a link</span>
			<Link className={getThemeClassname('NotFound__link', dark)} to="/">
				home
			</Link>
			?
		</h2>
	</div>
);

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

export default withTitle('Not Found')(connect(mapStateToProps)(NotFound));
