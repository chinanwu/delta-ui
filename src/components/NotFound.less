.NotFound {
	height: 100%;
	width: 100%;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
}

.NotFound__link {
	display: flex;
}

.NotFound__linkWords::after {
	content: '\00a0';
}
