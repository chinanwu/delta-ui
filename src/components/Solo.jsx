import FocusTrap from 'focus-trap-react';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Confetti from 'react-confetti';

import {
	aBtn,
	backspaceBtn,
	enterBtn,
	leftArrowBtn,
	rightArrowBtn,
	tabBtn,
	zBtn,
} from '../constants/Keycodes';
import { getFetch } from '../functions/FetchFunctions';
import formatCentisecondsTimer from '../functions/formatCentisecondsTimer';
import getThemeClassname from '../functions/getThemeClassname';
import isOneOff from '../functions/isOneOff';
import { applyFrom, applyTo, applyGame } from '../thunk/GameThunk.jsx';

import Loading from './Loading.jsx';
import HintButton from './HintButton.jsx';
import StealthForm from './StealthForm.jsx';
import ThemeToggle from './ThemeToggle.jsx';
import WinModal from './WinModal.jsx';

import './Solo.less';

const maxLen = 4;
const regex = /([a-zA-Z])/g;

export const Solo = ({
	dark,
	from,
	to,
	onChangeFrom,
	onChangeTo,
	onChangeGame,
}) => {
	const [win, setWin] = useState(false);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState('');

	const [numHints, setNumHints] = useState(3);
	const [hint, setHint] = useState(null);
	const [isHintExpanded, setIsHintExpanded] = useState(false);
	const [showHintInHistory, setShowHintInHistory] = useState(false);

	const [history, setHistory] = useState([]);
	const [guessVals, setGuessVals] = useState([]);
	const inputRefs = [useRef(null), useRef(null), useRef(null), useRef(null)];
	const historyBottomRef = useRef(null);

	const [timer, setTimer] = useState(0);

	const [solution, setSolution] = useState(null);
	const [score, setScore] = useState(0);

	useEffect(() => {
		let interval = null;
		if (!win) {
			interval = setInterval(() => {
				setTimer(seconds => seconds + 1);
			}, 10);
		}
		return () => clearInterval(interval);
	}, [timer, win]);

	useEffect(() => {
		document.title = 'Solo - Delta';
	}, []);

	useEffect(() => {
		if (!from && !to) {
			setLoading(true);
			getFetch('/api/v1/words')
				.then(res => {
					if (res.from && res.to) {
						onChangeGame({ from: res.from, to: res.to });
						setHistory([res.from]);
						setGuessVals(res.from.match(regex));
					} else {
						// Error is for guessing, need another way to log this
						// setError('Something went wrong grabbing game words!');
					}
				})
				.then(() => setLoading(false))
				.catch(err => console.log(err));
		} else {
			setHistory([from]);
			setGuessVals(from.match(regex));
		}
	}, [
		from,
		to,
		onChangeFrom,
		onChangeTo,
		setLoading,
		setHistory,
		setGuessVals,
	]);

	useEffect(() => {
		historyBottomRef.current.scrollIntoView({
			behavior: 'smooth',
		});
	}, [history]);

	const handleNewClick = useCallback(() => {
		setWin(false);
		setNumHints(3);
		setHint(null);
		setIsHintExpanded(false);
		setShowHintInHistory(false);
		setError(null);
		setTimer(0);

		setSolution(null);
		inputRefs.forEach(inputRef => {
			inputRef.current.value = '';
		});
		setLoading(true);
		getFetch('/api/v1/words')
			.then(res => {
				if (res.from && res.to) {
					onChangeGame({ from: res.from, to: res.to });
					setTimer(0);
				} else {
					// Error is for guessing, need another way to log this
					// setError('Something went wrong grabbing game words!');
				}
			})
			.then(() => setLoading(false));
	}, [
		onChangeGame,
		inputRefs,
		setWin,
		setError,
		setHint,
		setNumHints,
		setIsHintExpanded,
		setShowHintInHistory,
		setSolution,
		setLoading,
		setTimer,
	]);

	const handleHintClick = useCallback(() => {
		setLoading(true);
		getFetch(`/api/v1/hint?from=${history[history.length - 1]}&to=${to}`)
			.then(res => {
				if (res.hint && (res.numLeft === 0 || res.numLeft !== null)) {
					setNumHints(numHints => numHints - 1);
					// num left would be all words after last history word, includes to
					// e.g. heat -> cold, hint would be head, and numLeft 4 [head, held, hold, cold]
					setHint({ word: res.hint, numLeft: res.numLeft });
					setShowHintInHistory(true);
				} else {
					// Error is for guessing, need another way to log this
					// setError('Something went wrong grabbing game words!');
				}
			})
			.then(() => setLoading(false));
	}, [
		numHints,
		history,
		setNumHints,
		setHint,
		setShowHintInHistory,
		setLoading,
	]);

	const handleGetSoln = useCallback(() => {
		setLoading(true);
		getFetch(`/api/v1/solve?from=${from}&to=${to}`)
			.then(res => {
				if (res.solution) {
					setSolution(res.solution);
				} else {
					// Error is for guessing, need another way to log this
					// setError('Something went wrong grabbing game words!');
				}
			})
			.then(() => setLoading(false));
	}, [from, to, setLoading, setSolution]);

	const handleExpandHint = useCallback(
		expand => {
			setIsHintExpanded(expand);
		},
		[setIsHintExpanded]
	);

	const handleSubmitEdit = useCallback(
		(nextFrom, nextTo) => {
			setNumHints(3);
			setHint(null);
			setIsHintExpanded(false);
			setShowHintInHistory(false);
			setError(null);
			onChangeGame({ from: nextFrom, to: nextTo });
			setTimer(0);
		},
		[
			setNumHints,
			setHint,
			setIsHintExpanded,
			setShowHintInHistory,
			setError,
			setTimer,
			onChangeGame,
		]
	);

	const handleKeyDown = useCallback(
		event => {
			if (event && event.target && event.keyCode && event.key) {
				const i = parseInt(event.target.dataset.id);
				const next = (i + 1) % maxLen;
				const prev = i - 1 < 0 ? maxLen - 1 : i - 1;

				if (event.keyCode >= aBtn && event.keyCode <= zBtn) {
					event.preventDefault();
					guessVals[i] = event.key;
					inputRefs[i].current.value = event.key;
					if (next >= i) {
						inputRefs[next].current.focus();
						inputRefs[next].current.select();
					}
				} else {
					switch (event.keyCode) {
						case enterBtn:
							event.preventDefault();
							handleEnterClick();
							inputRefs[0].current.focus();
							break;
						case backspaceBtn:
							event.preventDefault();
							inputRefs[i].current.value = '';
							if (prev <= i) {
								inputRefs[prev].current.focus();
								inputRefs[prev].current.select();
							}
							break;
						case rightArrowBtn:
							inputRefs[next].current.focus();
							inputRefs[next].current.select();
							break;
						case leftArrowBtn:
							inputRefs[prev].current.focus();
							inputRefs[prev].current.select();
							break;
						case tabBtn:
							//do nothing
							break;
						default:
							event.preventDefault();
							break;
					}
				}
			}
		},
		[inputRefs]
	);

	const handleEnterClick = useCallback(() => {
		const guess = guessVals.join('');
		inputRefs.forEach(inputRef => {
			inputRef.current.value = '';
		});

		if (isOneOff(guess, history[history.length - 1])) {
			if (guess === to) {
				setLoading(true);
				getFetch(
					`/api/v1/score?from=${from}&to=${to}&time=${timer}&hintsUsed=${3 -
						numHints}&solution=${[...history, guess]}`
				)
					.then(res => {
						if (res) {
							setScore(res.score);
							setSolution(res.optimalSolution);
							setWin(true);
							setError(null);
							setHistory(history => [...history, guess]);
						}
					})
					.then(() => setLoading(false));
			} else {
				getFetch(`/api/v1/validate?word=${guess}`).then(res => {
					if (res.success) {
						setHistory(history => [...history, guess]);
						setError(null);
						setShowHintInHistory(false);
					} else {
						setGuessVals(history[history.length - 1].match(regex));
						setError("Word entered isn't a real word");
					}
				});
			}
		} else {
			setGuessVals(history[history.length - 1].match(regex));
			setError('Word must be one letter off from previous word');
		}
	}, [
		from,
		to,
		timer,
		numHints,
		guessVals,
		history,
		setScore,
		setSolution,
		setWin,
		setError,
		setHistory,
		setLoading,
		setShowHintInHistory,
		setGuessVals,
		setTimer,
	]);

	const makeMysteryStepListItems = num => {
		let all = [];
		for (let i = 0; i < num; i++) {
			all[i] = (
				<li
					className="Solo__historyHintMysteryLi"
					key={`Solo__historyHintMystery--${i}`}
				>
					<div className="Solo__historyHintMystery">?</div>
				</li>
			);
		}
		return all;
	};

	return (
		<div className={getThemeClassname('Solo', dark)}>
			<ThemeToggle />

			<StealthForm
				from={from}
				to={to}
				dark={dark}
				onChange={handleSubmitEdit}
			/>

			<div className="Solo__timer" role="timer">
				{formatCentisecondsTimer(timer)}
			</div>

			<div className={getThemeClassname('Solo__game', dark)}>
				<div className="Solo__history" aria-labelledby="soloHistory">
					<h3 id="soloHistory" className="Solo__historyLabel">
						History
					</h3>
					<ul
						className={
							getThemeClassname('Solo__historyList', dark) +
							(win ? ' Solo__historyList--hideScroll' : '')
						}
						aria-live="polite"
					>
						{history.map((item, i) => (
							<li key={`Solo__historyItem--${i}`}>{item}</li>
						))}
						<li ref={historyBottomRef} />

						{showHintInHistory && hint ? (
							hint.word !== to ? (
								<>
									<li
										className={getThemeClassname('Solo__historyHintWord', dark)}
									>
										{hint.word}
									</li>
									{makeMysteryStepListItems(hint.numLeft - 2)}
									<li>{to}</li>
								</>
							) : (
								<>
									<li className="Solo__historyHintWord">{hint.word}</li>
								</>
							)
						) : null}
					</ul>
				</div>

				<div>
					<HintButton
						id="soloHintBtn"
						numHints={numHints}
						isExpanded={isHintExpanded}
						dark={dark}
						btnText={`Get a Hint: ${numHints}`}
						ariaLabelledBy="soloHintHeader"
						giveSolution={true}
						onClick={handleHintClick}
						onSolnClick={handleGetSoln}
						onExpandChange={handleExpandHint}
					>
						<h3 id="soloHintHeader">Hint: </h3>
						<p className="Solo__hint">
							Recommended next word: {hint ? hint.word : ''}
						</p>
						<p className="Solo__hint">
							Estimated number of words left: {hint ? hint.numLeft : ''}
						</p>
					</HintButton>

					<div className="Solo__guessContainer">
						<h3 id="soloGuessLabel" className="Solo__guessLabel">
							Next Word:
						</h3>
						<div className="Solo__guessInputs">
							<div>
								{guessVals.map((val, i) => (
									<input
										id={'soloGuessInput-' + i}
										ref={inputRefs[i]}
										data-id={i}
										className={getThemeClassname('Solo__guessInput', dark)}
										type="text"
										maxLength={1}
										placeholder={val}
										aria-labelledby="soloGuessLabel"
										aria-invalid={error}
										onKeyDown={handleKeyDown}
										key={'Solo__guessInput-' + i}
									/>
								))}
							</div>

							<button
								id="soloEnterBtn"
								className="Solo__enterBtn"
								onClick={handleEnterClick}
							>
								Enter
							</button>
						</div>

						<div className={getThemeClassname('Solo__gameError', dark)}>
							{error}
						</div>
					</div>
				</div>
			</div>

			{loading && createPortal(<Loading />, document.body)}
			{win &&
				createPortal(
					<FocusTrap>
						<div className="Solo__win">
							<Confetti number={50} recycle={false} />
							<WinModal
								dark={dark}
								from={from}
								to={to}
								playerSoln={history}
								timer={timer}
								hintsUsed={3 - numHints}
								score={score}
								solution={solution}
								onNewGame={handleNewClick}
							/>
						</div>
					</FocusTrap>,
					document.body
				)}
			{solution &&
				createPortal(
					<div className="Solo__solution">
						<div className="Solo__solutionContent">
							<h2 className="Solo__solutionHeader">Solution</h2>

							<ul className="Solo__solutionList">
								{solution.map((step, i) => (
									<li key={`Solo__solution-${i}`}>{step}</li>
								))}
							</ul>

							<button
								id="soloWinNewGame"
								className="Solo__winBtn"
								aria-label="New Game"
								onClick={handleNewClick}
							>
								New Game
							</button>
						</div>
					</div>,
					document.body
				)}

			<div className="Solo__easterEgg">This is an incredibly tiny screen</div>
		</div>
	);
};

Solo.propTypes = {
	dark: PropTypes.bool,
	from: PropTypes.string,
	to: PropTypes.string,
	onChangeFrom: PropTypes.func,
	onChangeTo: PropTypes.func,
	onChangeGame: PropTypes.func,
};

export const mapStateToProps = ({
	game: {
		solo: { from, to },
	},
	theme: { dark },
}) => ({
	dark,
	from,
	to,
});

const mapDispatchToProps = {
	onChangeFrom: applyFrom,
	onChangeTo: applyTo,
	onChangeGame: applyGame,
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Solo);

// Fun Ideas:
// - Reverse button. Work from To -> From

// TODO:
// - Check for accessibility - Ongoing, forever and ever
// 			- Missing aria?
// - Error handling, need to figure out best way to deal with errors from the API
// - Similarly, should I improve the response check?
// - Loading component needs to be made
// - Add styling specific to each platform (e.g. moz, etc.)
// - Solution Modal dark mode
// - Constant Error strings instead of having it typed out over and over
// - Error causes input box to be red maybe?

// https://wireframe.cc/mNTM9B
