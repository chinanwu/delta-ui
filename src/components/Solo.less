@import '../style/styles';
@import '../style/button';

.Solo--dark {
	@dark();
}

.Solo__words {
	display: flex;
	justify-content: center;
	align-items: center;
	padding-top: 3rem;
}

.Solo__btns {
	display: flex;
	justify-content: center;
}

.Solo__btn {
	@smBtn();
	background-color: @black;
	color: white;

	&:hover {
		background-color: @blackHover;
	}

	&:active {
		background-color: @blackActive;
	}
}

.Solo__editor {
	animation: slideDown 1s;
	margin-top: 1rem;
	width: 100%;
	display: flex;
	justify-content: center;
}

@keyframes slideDown {
	from {
		transform: translateY(-1em);
		opacity: 0;
		visibility: hidden;
	}
	to {
		transform: translateY(0);
		opacity: 1;
		visibility: visible;
	}
}

.Solo__editor--hidden {
	display: none;
}

.Solo__editorContent {
	display: flex;
	flex-direction: column;
	justify-content: center;
	border: 1px solid black;
	padding: 1rem;
}

.Solo__editorBtns {
	display: flex;
	justify-content: center;
	margin-top: 1rem;
}

.Solo__editorBtn {
	@smBtn();
	background-color: @green;

	&:hover {
		background-color: @greenHover;
	}

	&:active {
		background-color: @greenActive;
	}
}

@media only screen and (max-width: 375px) {
	.Solo__btn,
	.Solo__editor,
	.Solo__editorBtn {
		font-size: 125%;
	}

	.Solo__editor,
	.Solo__editor--hiding {
		width: initial;
		margin-left: 2rem;
		margin-right: 2rem;
	}

	.Solo__editorContent {
		padding: 2rem;
	}

	.Solo__editorForms {
		display: flex;
		flex-direction: column;
	}

	.Solo__editorForm {
		display: flex;
		justify-content: space-between;
		align-items: center;
	}

	.Solo__editorLabel,
	.Solo__EditorInput {
		font-size: 1.8rem;
	}

	.Solo__EditorInput {
		width: 15rem;
	}

	.Solo__editor--center {
		text-align: center;
	}
}
