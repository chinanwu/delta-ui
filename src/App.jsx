import React from 'react';
import { Switch, Route } from 'react-router-dom';

import Home from './components/Home.jsx';
import Solo from './components/Solo.jsx';
import NotFound from './components/NotFound.jsx';

import './App.less';

export const App = () => (
	<Switch>
		<Route exact path="/" component={Home} />
		<Route path="/solo" component={Solo} />
		<Route component={NotFound} />
	</Switch>
);

export default App;
