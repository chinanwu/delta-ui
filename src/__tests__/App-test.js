import { shallow, mount } from 'enzyme';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { App } from '../App.jsx';
import Home from '../components/Home.jsx';
import NotFound from '../components/NotFound.jsx';
import Solo from '../components/Solo/Solo.jsx';

jest.unmock('../App.jsx');
jest.unmock('../components/NotFound.jsx'); // I have truly NO idea why I must do this to make the test pass

describe('App component', () => {
	describe('rendering', () => {
		it('renders App', () => {
			const wrapper = shallow(<App />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('routing', () => {
		it('routes to home', () => {
			const wrapper = mount(
				<MemoryRouter initialEntries={['/']} initialIndex={0}>
					<App />
				</MemoryRouter>
			);
			expect(wrapper.find(Home)).toHaveLength(1);
		});

		it('routes to solo', () => {
			const wrapper = mount(
				<MemoryRouter initialEntries={['/solo']} initialIndex={0}>
					<App />
				</MemoryRouter>
			);
			expect(wrapper.find(Home)).toHaveLength(0);
			expect(wrapper.find(Solo)).toHaveLength(1);
			expect(wrapper.find(NotFound)).toHaveLength(0);
		});

		it('routes to not found', () => {
			const wrapper = mount(
				<MemoryRouter initialEntries={['/fakeurllol']}>
					<App />
				</MemoryRouter>
			);
			expect(wrapper.find(Home)).toHaveLength(0);
			expect(wrapper.find(Solo)).toHaveLength(0);
			expect(wrapper.find(NotFound)).toHaveLength(1);
		});
	});
});
