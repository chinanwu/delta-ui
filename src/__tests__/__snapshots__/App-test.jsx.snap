// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component rendering renders App 1`] = `
<Switch>
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [MockFunction],
        "apply": [MockFunction],
        "bind": [MockFunction],
        "call": [MockFunction],
        "compare": null,
        "displayName": "Connect(Home)",
        "type": [MockFunction],
      }
    }
    exact={true}
    path="/"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [MockFunction],
        "apply": [MockFunction],
        "bind": [MockFunction],
        "call": [MockFunction],
        "compare": null,
        "displayName": "Connect(Solo)",
        "type": [MockFunction],
      }
    }
    path="/solo"
  />
  <Route
    component={[Function]}
  />
</Switch>
`;
